TempoWorld/Source/TempoWorld/Private/TempoWorldStateServiceSubsystem.cpp

```diff
@@ -0,0 +1,296 @@
+		return;
+	}
+
+	Response = GetActorState(Actors[0], GetWorld());
+	ResponseContinuation.ExecuteIfBound(Response, grpc::Status_OK);
+}
+
+void UTempoWorldStateServiceSubsystem::GetCurrentActorStatesNear(const TempoWorld::ActorStatesNearRequest& Request, const TResponseDelegate<TempoWorld::ActorStates>& ResponseContinuation)
+{
+	ActorStates Response;
+
+	const TArray<AActor*> Actors = GetMatchingActors(GetWorld(), Request);
+
+	for (const AActor* Actor : Actors)
+	{
+		*Response.add_actor_states() = GetActorState(Actor, GetWorld());
+	}
+
+	ResponseContinuation.ExecuteIfBound(Response, grpc::Status_OK);
+}
+
+void UTempoWorldStateServiceSubsystem::StreamActorState(const TempoWorld::ActorStateRequest& Request, const TResponseDelegate<TempoWorld::ActorState>& ResponseContinuation)
+{
+	PendingActorStateRequests.FindOrAdd(Request).Add(ResponseContinuation);
+}
+
+void UTempoWorldStateServiceSubsystem::StreamActorStatesNear(const TempoWorld::ActorStatesNearRequest& Request, const TResponseDelegate<TempoWorld::ActorStates>& ResponseContinuation)
+{
+	PendingActorStatesNearRequests.FindOrAdd(Request).Add(ResponseContinuation);
+}
+
+void UTempoWorldStateServiceSubsystem::StreamOverlapEvents(const OverlapEventRequest& Request, const TResponseDelegate<OverlapEventResponse>& ResponseContinuation)
+{
+	for (TActorIterator<AActor> ActorIt(GetWorld()); ActorIt; ++ActorIt)
+	{
+		if (ActorIt->IsHidden())
+		{
+			continue;
+		}
+		AActor* Actor = *ActorIt;
+		if (Actor->GetActorNameOrLabel().Equals(UTF8_TO_TCHAR(Request.actor_name().c_str()), ESearchCase::IgnoreCase))
+		{
+			PendingOverlapRequests.FindOrAdd(Actor->GetActorNameOrLabel()).Add(ResponseContinuation);
+			Actor->OnActorBeginOverlap.AddDynamic(this, &UTempoWorldStateServiceSubsystem::UTempoWorldStateServiceSubsystem::OnActorOverlap);
+		}
+	}
+}
+
+void UTempoWorldStateServiceSubsystem::OnActorOverlap(AActor* OverlappedActor, AActor* OtherActor)
+{
+	if (const auto ResponseContinuations = PendingOverlapRequests.Find(OverlappedActor->GetActorNameOrLabel()))
+	{
+		OverlapEventResponse Response;
+		Response.set_overlapped_actor_name(TCHAR_TO_UTF8(*OverlappedActor->GetActorNameOrLabel()));
+		Response.set_other_actor_name(TCHAR_TO_UTF8(*OtherActor->GetActorNameOrLabel()));
+		if (const ATempoGameMode* TempoGameMode = Cast<ATempoGameMode>(UGameplayStatics::GetGameMode(this)))
+		{
+			if (const IActorClassificationInterface* ActorClassifier = TempoGameMode->GetActorClassifier())
+			{
+				Response.set_other_actor_type(TCHAR_TO_UTF8(*ActorClassifier->GetActorClassification(OtherActor).ToString()));
+			}
+		}
+
+		for (const auto& ResponseContinuation : *ResponseContinuations)
+		{
+			ResponseContinuation.ExecuteIfBound(Response, grpc::Status_OK);
+		}
+		PendingOverlapRequests.Remove(OverlappedActor->GetActorNameOrLabel());
+		OverlappedActor->OnActorBeginOverlap.RemoveAll(this);
```

Comment:
Is your intent just to get the first overlap event for a given actor?  If so, wouldn't we want to also repeat this call for `OtherActor`?

Or, do we want to get subsequent overlap events for both?

Reply:
If we added this delegate as a result of a streaming API it will be re-added (and another pending request will be added to `PendingOverlapRequests`) immediately, as long as a client is still streaming it. I'm not sure what you mean by `wouldn't we want to also prevent OtherActor from getting further overlap events`? We're not listening for `OtherActor`'s overlap events, it's just the actor who overlapped the one we are listening for.

Reply:
Ahh, okay.  Looking again, I see we're only subscribing based on one name in the request.  (Cloudy-headed review due to post-conference cold.)